lib: os: printk hook cleanups

fabiobaltieri opened the pull request:
Make `_char_out` static, cleanup the console driver code.

Will follow up on `__stdout_hook_install` separately (https://github.com/zephyrproject-rtos/zephyr/pull/76234).

fabiobaltieri committed:
lib: os: make _char_out static

Looks like this was really meant to be static, there are explicit
getters and setters for it anyway.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri committed:
lib: os: add a header for printk hook functions

Add a zephyr/printk.h header for the __printk_hook functions, these are
currently manually declared by all console drivers for no good reason.

Move the documentation into the header and also unify the way that
console drivers call the function.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

andyross reviewed (dismissed):
Seems clean, and a reasonable step in the direction of squaring our weird knot of kernel output primitives.  We have printk, console and logging.  And most of them can be directed at others via asymmetric means, but not in a simple way.  In the longer term it might be nice if you could have a generic "handle character output" function and use it as any/all of:

* printk hook
* Console stdout hook
* Log backend

Likewise, the decision of "where to point logging" and "where to point printk" should use the same terminology and device/API naming.  Right now you can have CONFIG_LOG_PRINTK (means "implement printk in terms of logging") which makes the printk hook a noop.  Also CONFIG_LOG_MINIMAL (means "implement logging in terms of printk"), which then relies on someone else setting up printk.

And similarly, the printf() call[1] in hello_world often refuses to work on some platforms /configs, because the console isn't set up, as everyone using those platforms doesn't care as they all do logging or printk.

None of which has anything to do with this PR which seems clean.  But something to think about if you're in the space.

[1] It used to be a printk().  Not sure when it changed but I consistently and reliably trip over the distinction.

fabiobaltieri commented:
Thanks for the summary Andy - yes I bumped into this because I was looking into  figuring out the various interaction between logging with backends, logging minimal, logging with the shell backend etc... there's few combinations, not all of which necessarily lead to useful results. :-) And for the console it's odd that you can select multiple backends and they race for which one is the last setting the hook, was considering to convert the hooks into an iterable section, but then bumped into the early printk stuff as well, figured I'd just clean the headers for now.

This whole thing is a bit overwhelming for what it does.

fabiobaltieri commented on drivers/console/efi_console.c:
```diff
@@ -16,7 +16,7 @@
 #include <zephyr/init.h>
 #include <zephyr/sys/printk.h>
-
+#include <zephyr/printk-hooks.h>
```
I really just copied this from `zephyr/sys/libc-hooks.h`. Sholud I move it under `sys`? I'll move it under `sys/`.

andyross commented on include/zephyr/printk-hooks.h:
```diff
@@ -0,0 +1,29 @@
+ * routine that outputs one ASCII character at a time.
+ * @param fn putc routine to install
+ */
+void __printk_hook_install(int (*fn)(int c));
```
One nitpick:  if you're promoting this to an API a `typedef int (*printk_hook_fn_t)(int c);` might look cleaner than a bare function pointer.

fabiobaltieri replied:
Not a bad idea, added a patch for that.

fabiobaltieri commented:
moved the header under `sys/`

fabiobaltieri committed:
lib: os: use a typedef for __printk_hook_install argument

Use a typedef for the printk hook function argument.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
- added a typedef for the hook argument

fabiobaltieri committed:
lib: os: change __printk_get_hook return type to printk_hook_fn_t

Not sure why this was returning a void * but it seems to make sense to
return the correct type instead.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

keith-zephyr reviewed (approved):
Typo in the last commit message `insteadl`.

fabiobaltieri commented:
> Typo in the last commit message `insteadl`.

Thanks, fixed.
